pkg/corerp/frontend/controller/applications/graph_util.go

```diff
@@ -497,21 +491,55 @@ func connectionsFromAPIData(resource generated.GenericResource) []*corerpv202310
 		data := corerpv20231001preview.ConnectionProperties{}
 		err := toStronglyTypedData(connection, &data)
 		if err == nil {
+			sourceID, _ := findSourceResource(to.String(data.Source), allResources)
+
 			entries = append(entries, &corerpv20231001preview.ApplicationGraphConnection{
-				ID:        data.Source,
-				Direction: &dir,
+				ID:        to.Ptr(sourceID),
+				Direction: to.Ptr(dir),
 			})
 		}
 	}
 
 	// Produce a stable output
 	sort.Slice(entries, func(i, j int) bool {
-		return *entries[i].ID < *entries[j].ID
+		return to.String(entries[i].ID) < to.String(entries[j].ID)
 	})
 
 	return entries
 }
 
+// findSourceResource looks up resource id by using source string by the following steps:
+// 1. Immediately return the resource ID if the source is a valid resource ID.
+// 2. Parse the hostname from source and look up the hostname in the resource list if the source is a valid URL.
+// 3. Otherwise, return the original source string with false boolean value.
+func findSourceResource(source string, allResources []generated.GenericResource) (string, bool) {
+	// 1. Return the resource id if the source is a valid resource ID
+	id, err := resources.Parse(source)
+	if err == nil && id.IsResource() {
+		return id.String(), true
+	}
+
+	// 2. Parse hostname from source and look up hostname in resource list.
+	orig := source
+
+	// Add "//" to source to enable url.Parse to parse source correctly if the scheme is not given.
+	if !strings.Contains(source, "//") {
+		source = "//" + source
+	}
+
+	sourceURL, err := url.Parse(source)
+	if err == nil {
```

Comment:
nit: it would be more conventional invert this if:

```go
if err != nil {
  // 3. Return the original source string with false boolean value.
  return orig, false
}

Reply:
There is the case where I want to return original source if linear search cannot find the matched resource. So I chose this way, but I added comment on the code.

Reply:
no worries, just a nit

Reply:
Would we add a new ticket to ensure we don't have two resources in 2 different resource groups in the same env sharing the same name?